pdir/attr_category.py

```diff
@@ -261,10 +265,29 @@ def is_descriptor(obj):
                 AttrCategory.STATIC_METHOD,
                 AttrCategory.FUNCTION,
             )
+        elif isinstance(attr, classmethod):
+            return (
+                AttrCategory.DESCRIPTOR,
+                AttrCategory.CLASS_METHOD,
+                AttrCategory.FUNCTION,
+            )
         elif is_descriptor(attr):
             # Maybe add getsetdescriptor memberdescriptor in the future.
             return AttrCategory.DESCRIPTOR, AttrCategory.PROPERTY
         else:
-            # attr that is neither function nor class is a normal variable,
-            # and it's classified to property.
-            return AttrCategory.PROPERTY
+            # attr that is neither function nor class is a normal variable
+            # Possible cases: local variable (checking current frame),
+            # instance variable, class variable
+
+            # if we are checking current frame
+            if obj is dummy_obj:
```

Comment:
I don't believe there are more cases, example: module, namedtuple. There are many more, and they should all be `AttrCategory.PROPERTY`.

Reply:
I think the best way to implement it is to treat them as a fallback case(like before), which requires you to inspect class vars from parent classes(maybe by iterating over mro and check classes' `__dict__`).

Reply:
My opinion here is that modules as well as namedtuple can be considered as classes (see `_source` of namedtuple). If we really wish to treat them separately maybe we have to give a clear definition of their difference with user-defined class first. For example, a new category called "module variable"  could be applied to modules to replace the perhaps misleading "instance variable".
On the other hand, setting up so many categories isn't some quick fix, so fall back to a simple "property" for now might not be a bad idea.

Reply:
When I came to modify the code I felt that I've misunderstood what you've meant and I myself wasn't thinking very clearly yesterday. Below are two different ways how I can interpret your idea and corresponding ways to implement them.
#### What I thought you meant yesterday
For cases such as `module` or `namedtuple` some of their attributes in `dir` can not be categorized into `instance variable` or `class variable` and should be temporarily categorized into `property`. For example, if a module `m` has defined a variable `mv`, the current (this GitHub branch) behavior for `pdir(m)` is:
```
instance variable:
    mv
```
which to some extend is misleading. What should be expected is:
```
property:
    mv
```
And as I suggested maybe in future it'll be more clearly expressed as:
```
module variable:
    mv
```
If this was what you've meant, what should be done is to **judge the type of `obj`** in `get_attr_category` and goto different branch according to their type. A fallback in the last clause isn't very helping.
####  Another way to take it
Suppose we still have the `m` module, for the following class definition:
```
class A:
    fancy_module = m
```
the current behavior of `pdir(A())`  is:
```
class variable:
    fancy_module
```
in which `fancy_module` is (technically correctly but less helpfully) treated as a class variable. What should be expected is:
```
property:
    fancy_module
```
and maybe in future:
```
modules:
    fancy_module
```
If this was what you've meant, what should be done is to **judge the type of `attr`** in `get_attr_category` and goto different branch according to their type. Again, a fallback in the last clause isn't very helping.

At first I'm also a little uncomfortable by the fact that all remaining attributes are classified into 3 categories, but if you look at how attrs are gathered: `attrs = _getframe(1).f_locals if obj is dummy_obj else self._getattr()` it'll be clear that there're only three categories in principle.
* The first case ` _getframe(1).f_locals` represents local variable
* `self._getattr()` represents instance variable and class variable, because:
  * at the heart of `self._getattr()` is `get_dict_attr(self.obj, name)`. The `hasattr` is a patch which I'll discuss later in this comment.
  * in `get_dict_attr(self.obj, name)` what we do is to find the name in instance `__dict__` and class `__dict__`, so it's either some kind of instance variable or some kind of class variable.

#### To conclude
In my opinion both ways are okay and they can even be combined to provide better user experience. However I'm not sure what your idea is.

#### BTW: on the try...except `hasattr` patch.
After some digging I'd consider the patch not necessary or at least not explicit. This should probably be filed in another issue but as I've mentioned it in this comment let me briefly explain here. This patch is to solve cases when  `get_dict_attr(self.obj, name)` fail, and actually it fails in a certain case not so closely related to some specific packages (#1). See the following example:
```
class A:
    base = 1
class B(A): pass
pdir(B)
# exception catched and `attrs[name] = getattr(self.obj, name)` is called
```
The problem arises from inheritance and `__mro__`. If we pass a `class` as an `obj` to `pdir`, what happens in `attrs[name] = getattr(self.obj, name)` is that (code below) we tried to look up `base` in `B.__dict__` and failed and rather than go to his father we goto `B.__class__.__mro__` which in other words is `type(B).__mro__` and this certainly can not lead us to `base`, so in the end we have to raise the `AttributeError`.
```
def get_dict_attr(attr_obj, attr_name):
    for obj in [attr_obj] + list(attr_obj.__class__.__mro__):
        if hasattr(obj, '__dict__') and attr_name in obj.__dict__:
            return obj.__dict__[attr_name]
    raise AttributeError
```
A fix to the issue:
```
def get_dict_attr(attr_obj, attr_name):
    if inspect.isclass(attr_obj):
        obj_list = [attr_obj] + list(attr_obj.__mro__)
    else:
        obj_list = [attr_obj] + list(attr_obj.__class__.__mro__)
    for obj in obj_list:
        if hasattr(obj, '__dict__') and attr_name in obj.__dict__:
            return obj.__dict__[attr_name]
    raise AttributeError
```
And the try...catch in `self._getattr()` won't be necessary any more. The try...catch and `getattr` patch is not buggy, but can lead to weird behavior in some cases. For the following class definition:
```
class A():
    @classmethod
    @property
    def a(cls):
        return 'a'

class B(A): pass
```
`pdir(A)` produces:
```
...
descriptor:
    a: class classmethod with getter, classmethod(function) -> method
```
`pdir(B)` produces:
```
...
function:
    a: 
```
And my improved version of `get_dict_attr` uniformly produces:
```
...
descriptor:
    a: class classmethod with getter, classmethod(function) -> method
```
I'd probably include this refactoring in this PR.

Thanks for reading. I didn't expect it'll be this long.

Reply:
I haven't looked through the `self._getattr()` part, I think you probably want to open a separate issue. Now let me explain what I meant in previous comments.

In short, from end users' perspective, attributes in a namedtuple, or variables in a module(and many more), are properties and properties only. Categorizing them to `instance variable` or `class variable` is only gonna confuse people. 

For example,
```
>>> A = namedtuple('A', ['a'])
>>> isclass(A)
True
>>> A.a
<property object at 0x1102512c8>
```

Do people call `a` class variables? Usually they don't, and if we categorize it to `AttrCategory.CLASS_VARIABLE`, it confuses people.

The design principle for pdir2 is to not confuse people, and help them as much as we can. If we can't categorize things to a more detailed category, then a general one is also fine.

With that being said, I do think in your example
```
class A:
    a = 1         # regular var
    fancy_module = m   # a module
```
Categorizing it to `modules` would be super helpful, and is not confusing. This leads to the second point I want to make: **Why don't `CLASS_VARIABLE` and `INSTANCE_VARIABLE` fit very well in pdir2.**

If you take a look at existing categories, it's telling people: **how they should use these attributes**. `SLOT` is an exception, but we're not putting it as an independent category either. Now here comes `CLASS_VARIABLE` and `INSTANCE_VARIABLE`, they are telling people **where** those attributes are defined, without mentioning their usage. Since "location" and "functionality" are pretty much orthogonal to each other, it is very common that something has a location and a functionality. This brings a problem, because when presenting an attribute, pdir2 will pick one `display_group`, however, if an attribute has a `location` and a `functionality`, which one are you gonna choose? There are 2 options:

1. Prefer `functionality` to `location`. We will loose some consistency, people will say, "wait, `fancy_module` is a class variable, but it's not put into `CLASS_VARIABLE` like `a`, why?". This causes confusion.

2. The slot way, which is to display both. However we haven't find a good way to display yet. Putting `CLASS_VARIABLE` and `INSTANCE_VARIABLE` into brackets would look cumbersome, cause there could be a lot.

Since there's no perfect way handling it, I'm hesitate to add `CLASS_VARIABLE` and `INSTANCE_VARIABLE`. We want to have more categories, `module` would be a good one, because it fits into existing categories well and doesn't cause confusion.

Hope I've explained my points clear enough. Feel free to bring in more thoughts.

Reply:
Thanks for your reply. It's crystal clear.
On the first point, I'm not sure what "properties" means in your context. May I ask for a clear definition of it? If it means function decorated by `@property`, then I believe `pdir` already had sound codes to deal with the situation and this gh branch doesn't break them. For the example you gave:
```
>>> A = namedtuple('A', ['a'])
>>> pdir(A)
class variable:
    _fields, _source
special attribute:
    __class__, __doc__, __module__, __slots__
abstract class:
    __subclasshook__
arithmetic:
    __add__, __mul__, __rmul__
object customization:
    __format__, __hash__, __init__, __new__, __repr__, __sizeof__, __str__
rich comparison:
    __eq__, __ge__, __gt__, __le__, __lt__, __ne__
attribute access:
    __delattr__, __dir__, __getattribute__, __setattr__
class customization:
    __init_subclass__
container:
    __contains__, __getitem__, __iter__, __len__
pickle:
    __getnewargs__, __reduce__, __reduce_ex__
descriptor:
    a: @property with getter, Alias for field number 0
class method:
    _make: classmethod(function) -> method
function:
    _asdict: Return a new OrderedDict which maps field names to their values.
    _replace: Return a new A object replacing specified fields with new values
    count: T.count(value) -> integer -- return number of occurrences of value
    index: T.index(value, [start, [stop]]) -> integer -- return first index of value.
```
`a` still stays in `descriptor` with a tailing `@property`。
OTOH, I agree on the second point, though I've never thought about it before - instance variable and class variable aren't really good categories for the principle of `pdir`. One more confusing case is that functions and magic methods to some extent are all instance variables. If you decide not to add these categories, please close this PR.

Reply:
Sorry for not giving a clear definition. "properties" here has a very general meaning. Think about it, you have an `obj`, and basically there's only two ways you can use it:
```python
obj.something
obj.func(args)
```
I call the first one "properties", so it's really a gerneral term. To make it more formal: In pdir2, properties means something that is accessible via a dot `.`, but is not callable, and not categorized into other categories.

Reply:
So what I thought yesterday is basically correct then. On modules (and perhaps more cases) the problem is evident, but I still hold that for namedtuple it's okay.
IMHO for user-defined and package-defined classes using ideas like instance variable and class variable is not misleading if not helping. The problem usually arises from cases in "built-in" classes. There's a list of these classes in the module `types`. It's worth paying attention if instance variable and class variable are still going to be added to `pdir`.
And if another path in which "properties" are classified by their actual data type is chosen, then currently I see no position for the idea of instance variable and class variable. Need more reflection on this.